# Introduction

In this Project, we will learn on how to plot map for Toronto and impose our neighbourhood dataset with latitude and longitude information available to us.
We will use foursquare api to fetch most common venues for each neighborhood and then we will use this feature to form cluster of neighborhoods and place them on map.

### Importing libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!pip install folium
#!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

     |████████████████████████████████| 92kB 16.0MB/s eta 0:00:01
Libraries imported.


Following cell is hidden as it contains sensitive information about importing csv file from watson studio container to notebook.

In [2]:
# The code was removed by Watson Studio for sharing.

Dropping null valued rows from dataframe

In [3]:
torono_df = torono_df.dropna()
torono_df.head()

Postcode      borough                         neighbourhood  latitude  \
0      M1B  Scarborough                         Rouge,Malvern   43.8113   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   43.7878   
2      M1E  Scarborough       Guildwood,Morningside,West Hill   43.7678   
3      M1G  Scarborough                                Woburn   43.7712   
4      M1H  Scarborough                             Cedarbrae   43.7686   

   longitude  
0   -79.1930  
1   -79.1564  
2   -79.1866  
3   -79.2144  
4   -79.2389

Plotting map of Toronto as a base

In [4]:
geolocator = Nominatim(user_agent='myapp')
location = geolocator.geocode('Toronto, CA')
latitude = location.latitude
longitude = location.longitude

In [5]:

toronto_map = folium.Map(location=[latitude,longitude],zoom_start=13)
toronto_map

Plotting Neighbourhood data in existing Toronoto Map. As the traditional .format method was not working for label, i had to use popup method from folium for label creation.

In [227]:
for lat,lon, borough, neighbourhood in zip(torono_df['latitude'], torono_df['longitude'], torono_df['borough'], torono_df['neighbourhood']):
    #print("Borough - {}, Neighbourhood - {}, latitude - {}, longitude - {}".format(borough,neighbourhood,lat,lon))
    label = folium.Popup(neighbourhood,parse_html=True)
    
    folium.CircleMarker(
        location=[lat,lon],
        color = 'red',
        radius = 5,
        popup=label,
        parse_html = True).add_to(toronto_map)
display(toronto_map)

### Foursquare API Credentials and Version details

In [7]:
# The code was removed by Watson Studio for sharing.

I tested first for one neighbourhood rather than running for loop and calling api multiple times unnecessarily.

In [8]:
#for lat,lon, neighbourhood, Postcode in zip(torono_df['latitude'],torono_df['longitude'],torono_df['neighbourhood'],torono_df['borough']):

radius = 700
latitude  = torono_df.loc[torono_df['neighbourhood'] == 'Woburn']['latitude'].values[0]
longitude = torono_df.loc[torono_df['neighbourhood'] == 'Woburn']['longitude'].values[0]
LIMIT = 100
print(latitude)
print(longitude)
print(search_query)

43.7712
-79.2144
Woburn


In [10]:
# for each in search_query:
#     url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, each, radius, LIMIT)
#     url
#     results = requests.get(url).json()
#     print(results)
       
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url
results = requests.get(url).json()


In [22]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [59]:
nearby_venues = json_normalize(results['response']['groups'][0]['items'])
nearby_venues.head()
columns = ['venue.name','venue.categories','venue.location.lat','venue.location.lng']
nearby_venues = nearby_venues.loc[:,columns]
nearby_venues.head()
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type,axis = 1)
columns = [a.split('.')[-1] for a in nearby_venues.columns.values]
nearby_venues.columns = columns
#nearby_venues['venue.categories'][0][0]['name']

In [60]:
print(nearby_venues.head())
print(nearby_venues.shape)

                   name         categories        lat        lng
0             Starbucks        Coffee Shop  43.770037 -79.221156
1           Tim Hortons        Coffee Shop  43.770827 -79.223078
2      Convenience Plus  Convenience Store  43.771860 -79.209363
3  Heather Heights Park               Park  43.776594 -79.215052
4         Country Style        Coffee Shop  43.773716 -79.207027
(6, 4)


#### Since it is working fine, We can go ahead and construct for loop for all neighbourhoods.

In [95]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [96]:
test_df = getNearbyVenues(latitudes=torono_df['latitude'],longitudes=torono_df['longitude'],names=torono_df['neighbourhood'])

Rouge,Malvern
Highland Creek,Rouge Hill,Port Union
Guildwood,Morningside,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park,Ionview,Kennedy Park
Clairlea,Golden Mile,Oakridge
Cliffcrest,Cliffside,Scarborough Village West
Birch Cliff,Cliffside West
Dorset Park,Scarborough Town Centre,Wexford Heights
Maryvale,Wexford
Agincourt
Clarks Corners,Sullivan,Tam O'Shanter
Agincourt North,L'Amoreaux East,Milliken,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West,Riverdale
The Beaches West,Indi

Checking shape and content of dataframe

In [98]:
test_df.shape

(2267, 7)

In [117]:
test_df.head()

Neighborhood  Neighborhood Latitude  \
0                    Rouge,Malvern                43.8113   
1  Guildwood,Morningside,West Hill                43.7678   
2  Guildwood,Morningside,West Hill                43.7678   
3  Guildwood,Morningside,West Hill                43.7678   
4  Guildwood,Morningside,West Hill                43.7678   

   Neighborhood Longitude                             Venue  Venue Latitude  \
0                -79.1930  Canadian Appliance Source Whitby       43.808353   
1                -79.1866                       Chick-N-Joy       43.768752   
2                -79.1866              Little Caesars Pizza       43.769046   
3                -79.1866                         Bulk Barn       43.771342   
4                -79.1866                              LCBO       43.771462   

   Venue Longitude       Venue Category  
0       -79.191331         Home Service  
1       -79.187982  Fried Chicken Joint  
2       -79.184386          Pizza Place  
3       -79.184341    Food & Drink Shop  
4       -79.184384         Liquor Store

#### Let's check groupby count for each neighourhood 

In [128]:
test_df.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Agincourt                                                               5   
Agincourt North,L'Amoreaux East,Milliken,Steele...                      2   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                     10   
Alderwood,Long Branch                                                   7   
Bathurst Manor,Downsview North,Wilson Heights                           6   
Bayview Village                                                         4   
Bedford Park,Lawrence Manor East                                       24   
Berczy Park                                                            93   
Birch Cliff,Cliffside West                                              4   
Bloordale Gardens,Eringate,Markland Wood,Old Bu...                      9   
Brockton,Exhibition Place,Parkdale Village                             39   
Business Reply Mail Processing Centre 969 Eastern                      15   
CFB Toronto,Downsview East                                              6   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     57   
Cabbagetown,St. James Town                                             43   
Caledonia-Fairbanks                                                     9   
Cedarbrae                                                               4   
Central Bay Street                                                     89   
Chinatown,Grange Park,Kensington Market                                84   
Christie                                                               11   
Church and Wellesley                                                   88   
Clairlea,Golden Mile,Oakridge                                          10   
Clarks Corners,Sullivan,Tam O'Shanter                                  12   
Cliffcrest,Cliffside,Scarborough Village West                           9   
Cloverdale,Islington,Martin Grove,Princess Gard...                      7   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             24   
Davisville North                                                        6   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                      6   
Del Ray,Keelesdale,Mount Dennis,Silverthorn                             5   
Design Exchange,Toronto Dominion Centre                               100   
Don Mills North                                                         3   
Dorset Park,Scarborough Town Centre,Wexford Hei...                      2   
Dovercourt Village,Dufferin                                            18   
Downsview Central                                                       1   
Downsview Northwest                                                    18   
Downsview West                                                          2   
Downsview,North Park,Upwood Park                                        2   
East Birchmount Park,Ionview,Kennedy Park                              15   
East Toronto                                                            3   
Emery,Humberlea                                                         6   
Fairview,Henry Farm,Oriole                                             65   
First Canadian Place,Underground city                                 100   
Flemingdon Park,Don Mills South                                         3   
Forest Hill North,Forest Hill West                                      3   
Glencairn                                                               8   
Guildwood,Morningside,West Hill                                        31   
Harbord,University of Toronto                                          31   
Harbourfront                                                           25   
Harbourfront East,Toronto Islands,Union Station 

#### Let's see how many of these venues are actually unique

In [129]:
len(test_df['Venue Category'].unique())

264

We have 264 unique venues. Now we can go ahead and use One-Hot Enconding on dataframe for later use.

In [185]:
toronto_one_hot = pd.get_dummies(test_df[['Venue Category']], prefix="", prefix_sep="")
toronto_one_hot = toronto_one_hot.drop(columns=['Neighborhood'])
toronto_one_hot.head()

Accessories Store  Afghan Restaurant  Airport  American Restaurant  \
0                  0                  0        0                    0   
1                  0                  0        0                    0   
2                  0                  0        0                    0   
3                  0                  0        0                    0   
4                  0                  0        0                    0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0            0                    0                 0                   0   
1            0                    0                 0                   0   
2            0                    0                 0                   0   
3            0                    0                 0                   0   
4            0                    0                 0                   0   

   Auto Dealership  Auto Garage  BBQ Joint  Baby Store  Badminton Court  \
0                0            0          0           0                0   
1                0            0          0           0                0   
2                0            0          0           0                0   
3                0            0          0           0                0   
4                0            0          0           0                0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Basketball Court  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Stadium  Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0          0         0           0                   0   
1                   0          0         0           0                   0   
2                   0          0         0           0                   0   
3                   0          0         0           0                   0   
4                   0          0         0           0                   0   

   Bike Rental / Bike Share  Bistro  Bookstore  Boutique  \
0                         0       0          0         0   
1                         0       0          0         0   
2                         0       0          0         0   
3                         0       0          0         0   
4                         0       0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Buffet  \
0                     0               0        0                0       0   
1                     0               0        0                0       0   
2                     0               0        0                0       0   
3                     0               0        0                0       0   
4                     0               0        0                0       0   

   Building  Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
0         0             0              0         0            0         0   
1         0             0              0         0            0         0   
2         0             0              0         0            0         0   
3         0             0              0         0            0         0   
4         0             0              0         0            0         0   

   Business Service  Butcher  Cafeteria  Café  Camera Store  Candy Store  \
0                 0        0          0     0             0            0   
1                 0        0          0     0             0            0   
2                 0        0          0     0             0            0   
3                 0        0          0     0             0            0   
4                 0        0          0     0             0            0   

   Caribbean Res

#### Adding Neighbourhood column in the list of columns.

In [186]:
toronto_one_hot.insert(0,'Neighborhood',test_df['Neighborhood'])
toronto_one_hot.head()
# columns_fixing = [toronto_one_hot.columns[-1]] + list(toronto_one_hot.columns[:-1])
# toronto_one_hot = toronto_one_hot[columns_fixing]


Neighborhood  Accessories Store  Afghan Restaurant  \
0                    Rouge,Malvern                  0                  0   
1  Guildwood,Morningside,West Hill                  0                  0   
2  Guildwood,Morningside,West Hill                  0                  0   
3  Guildwood,Morningside,West Hill                  0                  0   
4  Guildwood,Morningside,West Hill                  0                  0   

   Airport  American Restaurant  Art Gallery  Arts & Crafts Store  \
0        0                    0            0                    0   
1        0                    0            0                    0   
2        0                    0            0                    0   
3        0                    0            0                    0   
4        0                    0            0                    0   

   Asian Restaurant  Athletics & Sports  Auto Dealership  Auto Garage  \
0                 0                   0                0            0   
1                 0                   0                0            0   
2                 0                   0                0            0   
3                 0                   0                0            0   
4                 0                   0                0            0   

   BBQ Joint  Baby Store  Badminton Court  Bagel Shop  Bakery  Bank  Bar  \
0          0           0                0           0       0     0    0   
1          0           0                0           0       0     0    0   
2          0           0                0           0       0     0    0   
3          0           0                0           0       0     0    0   
4          0           0                0           0       0     0    0   

   Baseball Field  Basketball Court  Basketball Stadium  Beach Bar  Beer Bar  \
0               0                 0                   0          0         0   
1               0                 0                   0          0         0   
2               0                 0                   0          0         0   
3               0                 0                   0          0         0   
4               0                 0                   0          0         0   

   Beer Store  Belgian Restaurant  Bike Rental / Bike Share  Bistro  \
0           0                   0                         0       0   
1           0                   0                         0       0   
2           0                   0                         0       0   
3           0                   0                         0       0   
4           0                   0                         0       0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bubble Tea Shop  Buffet  Building  Burger Joint  Burrito Place  Bus Line  \
0                0       0         0             0              0         0   
1                0       0         0             0              0         0   
2                0       0         0             0              0         0   
3                0       0         0             0              0         0   
4                0       0         0             0              0         0   

   Bus Station  Bus Stop  Business Service  Butcher  Cafeteria  Café  \
0            0         0                 0        0          0     0   
1            0         0                 0        0          0     0   
2            0         0                 0        0          0     0   
3            0         0                 0        0          0     0   
4            0         0                 0        0          0     0   


#### Now, Let's group dataframe by Neighbourhood column we just inserted and find mean for overall common places count

In [187]:
toronto_one_hot_grouped = toronto_one_hot.groupby('Neighborhood').mean().reset_index()

In [188]:
toronto_one_hot_grouped.head()

Neighborhood  Accessories Store  \
0                             Adelaide,King,Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...                0.0   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...                0.0   
4                              Alderwood,Long Branch                0.0   

   Afghan Restaurant  Airport  American Restaurant  Art Gallery  \
0                0.0      0.0                 0.03         0.01   
1                0.0      0.0                 0.00         0.00   
2                0.0      0.0                 0.00         0.00   
3                0.0      0.0                 0.00         0.00   
4                0.0      0.0                 0.00         0.00   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                  0.0              0.02                 0.0              0.0   
1                  0.0              0.00                 0.0              0.0   
2                  0.0              0.00                 0.0              0.0   
3                  0.0              0.00                 0.0              0.0   
4                  0.0              0.00                 0.0              0.0   

   Auto Garage  BBQ Joint  Baby Store  Badminton Court  Bagel Shop  Bakery  \
0          0.0        0.0         0.0              0.0         0.0    0.02   
1          0.0        0.0         0.0              0.2         0.0    0.00   
2          0.0        0.0         0.0              0.0         0.0    0.00   
3          0.0        0.0         0.0              0.0         0.0    0.00   
4          0.0        0.0         0.0              0.0         0.0    0.00   

   Bank   Bar  Baseball Field  Basketball Court  Basketball Stadium  \
0   0.0  0.03             0.0               0.0                 0.0   
1   0.0  0.00             0.0               0.0                 0.0   
2   0.0  0.00             0.0               0.0                 0.0   
3   0.0  0.00             0.0               0.0                 0.0   
4   0.0  0.00             0.0               0.0                 0.0   

   Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  \
0        0.0      0.01         0.0                 0.0   
1        0.0      0.00         0.0                 0.0   
2        0.0      0.00         0.0                 0.0   
3        0.0      0.00         0.1                 0.0   
4        0.0      0.00         0.0                 0.0   

   Bike Rental / Bike Share  Bistro  Bookstore  Boutique  \
0                       0.0     0.0       0.02       0.0   
1                       0.0     0.0       0.00       0.0   
2                       0.0     0.0       0.00       0.0   
3                       0.0     0.0       0.00       0.0   
4                       0.0     0.0       0.00       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Buffet  \
0                  0.01            0.02      0.0              0.0     0.0   
1                  0.00            0.20      0.0              0.0     0.0   
2                  0.00            0.00      0.0              0.0     0.0   
3                  0.00            0.00      0.0              0.0     0.0   
4                  0.00            0.00      0.0              0.0     0.0   

   Building  Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
0      0.01          0.02           0.01       0.0          0.0       0.0   
1      0.00          0.00           0.00       0.0          0.0       0.0   
2      0.00          0.00           0.00       0.0          0.0       0.0   
3      0.00          0.00           0.00       0.0          0.0       0.0   
4      0.00          0.00           0.00       0.0          0.0       0.0   

   Business Service  Butcher  Cafeteria  Café  Camera Store  Candy Store  \
0               0.0      0.0        0.0  0.06           0.0          0.0   
1              

#### Finding Top 3 venues from above grouped data by sorting them in decending order such as top 3 common places are chosen.

In [213]:
no_of_top_venues = 3
for each_nhood in toronto_one_hot_grouped['Neighborhood']:
    print("------------------{}------------------".format(each_nhood))
    
    temp_top = toronto_one_hot_grouped[toronto_one_hot_grouped['Neighborhood'] == each_nhood].T.reset_index()
    temp_top.columns = ['Venue','Value']
    temp_top = temp_top[1:]
    temp_top['Value'] = temp_top['Value'].astype(float)
    temp_top = temp_top.sort_values(['Value'],ascending=False).reset_index().drop('index',axis = 1).head(no_of_top_venues)
    print(temp_top)

------------------Adelaide,King,Richmond------------------
         Venue  Value
0         Café   0.06
1  Coffee Shop   0.06
2   Restaurant   0.05
------------------Agincourt------------------
                       Venue  Value
0  Latin American Restaurant    0.2
1             Breakfast Spot    0.2
2               Skating Rink    0.2
------------------Agincourt North,L'Amoreaux East,Milliken,Steeles East------------------
                Venue  Value
0    Sushi Restaurant    0.5
1            Pharmacy    0.5
2  Miscellaneous Shop    0.0
------------------Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown------------------
            Venue  Value
0   Grocery Store    0.2
1     Pizza Place    0.1
2  Hardware Store    0.1
------------------Alderwood,Long Branch------------------
               Venue     Value
0        Pizza Place  0.142857
1  Convenience Store  0.142857
2     Sandwich Place  0.142857
------------------Bathurst Manor,Dow

#### Let's List Top 7 common places in each neighbourhood and save it in a new dataframe

In [214]:
num_top_venues = 7

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

In [215]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [216]:
# create a new dataframe
tor_nhoods_venues_sorted = pd.DataFrame(columns=columns)
tor_nhoods_venues_sorted['Neighborhood'] = toronto_one_hot_grouped['Neighborhood']

for ind in np.arange(toronto_one_hot_grouped.shape[0]):
    tor_nhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_one_hot_grouped.iloc[ind, :], num_top_venues)

tor_nhoods_venues_sorted.dtypes

Neighborhood              object
1st Most Common Venue     object
2nd Most Common Venue     object
3rd Most Common Venue     object
4th Most Common Venue     object
5th Most Common Venue     object
6th Most Common Venue     object
7th Most Common Venue     object
8th Most Common Venue     object
9th Most Common Venue     object
10th Most Common Venue    object
dtype: object

#### Forming cluster and applying dataset to KMeans Object

In [217]:
no_of_cluster = 7
tor_nhoods_venues_sorted_cluster = toronto_one_hot_grouped.drop('Neighborhood',1)
kmeans= KMeans(n_clusters=5, random_state=0).fit(tor_nhoods_venues_sorted_cluster)
kmeans.labels_[0:10] 

array([2, 2, 2, 1, 2, 2, 2, 2, 2, 2], dtype=int32)

In [218]:
tor_nhoods_venues_sorted.insert(0,'Cluster Labels', kmeans.labels_)
torono_df_merged_final = torono_df
torono_df_merged_final = torono_df_merged_final.join(tor_nhoods_venues_sorted.set_index('Neighborhood'), on='neighbourhood')
torono_df_merged_final.head()

Postcode      borough                         neighbourhood  latitude  \
0      M1B  Scarborough                         Rouge,Malvern   43.8113   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union   43.7878   
2      M1E  Scarborough       Guildwood,Morningside,West Hill   43.7678   
3      M1G  Scarborough                                Woburn   43.7712   
4      M1H  Scarborough                             Cedarbrae   43.7686   

   longitude  Cluster Labels       1st Most Common Venue  \
0   -79.1930             4.0                Home Service   
1   -79.1564             NaN                         NaN   
2   -79.1866             2.0                 Pizza Place   
3   -79.2144             2.0           Convenience Store   
4   -79.2389             2.0  Construction & Landscaping   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop           Flower Shop           Flea Market   
1                   NaN                   NaN                   NaN   
2  Fast Food Restaurant     Electronics Store           Coffee Shop   
3     Korean Restaurant   Dumpling Restaurant           Flower Shop   
4                Lounge                 Trail           Gaming Cafe   

  5th Most Common Venue 6th Most Common Venue       7th Most Common Venue  \
0           Fish Market     Fish & Chips Shop  Financial or Legal Service   
1                   NaN                   NaN                         NaN   
2        Discount Store    Mexican Restaurant                   Juice Bar   
3           Flea Market           Fish Market           Fish & Chips Shop   
4           Yoga Studio        Farmers Market           Electronics Store   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                       Field  Fast Food Restaurant         Farmers Market  
1                         NaN                   NaN                    NaN  
2                 Supermarket   Fried Chicken Joint             Beer Store  
3  Financial or Legal Service                 Field   Fast Food Restaurant  
4        Ethiopian Restaurant           Event Space     Falafel Restaurant

#### Dropping rows with NaN values

In [225]:
torono_df_merged_final = torono_df_merged_final.dropna()

In [226]:
# create map


geolocator = Nominatim(user_agent='myapp',timeout=5)
location = geolocator.geocode('Toronto, CA')
latitude = location.latitude
longitude = location.longitude
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(no_of_cluster)
ys = [i + x + (i*x)**2 for i in range(no_of_cluster)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(torono_df_merged_final['latitude'], torono_df_merged_final['longitude'], torono_df_merged_final['neighbourhood'], torono_df_merged_final['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters